TykTechnologies/tyk-pump
# TT-5078 Feature/ Prometheus custom metrics

## Description

Added a new feature for Prometheus pump to allow users to create custom metrics.

How does it works?
Now you can add multiple custom_metrics in prometheus meta specifying the type of the metric and which labels / values is it going to use. Docs: https://github.com/TykTechnologies/tyk-pump/tree/feature/prometheus-custom-metrics#custom-prometheus-metrics

Example:
```
"prometheus": {
  "type": "prometheus",
	"meta": {
		"listen_address": "localhost:9090",
		"path": "/metrics",
        "custom_metrics":[
          {
            "name":"tyk_custom_http_status_per_api_name",
            "description":"This is a custom counter",
            "metric_type":"counter",
            "labels":["response_code","api_name"]
          }
        ]
	}
},
```

So in this way, you can aggregate the metrics the way you want and expose them in Prometheus.



## Related Issue




https://tyktech.atlassian.net/browse/TT-5078
## Motivation and Context

https://tyktech.atlassian.net/browse/TT-5078
## How This Has Been Tested


Added some unit tests.

In order to manually test it you need to:
- Set Prometheus pump with the description config.
- run Tyk Pump - send some traffic.
- hit your Prometheus pump /metrics configured endpoint, you should see your custom metric there.
- play  with different custom metric combination.


## Screenshots (if appropriate)

## Types of changes


## Checklist


      fork, don't request your `master`!
      *your branch* off *our latest `master`*.
